README.md
# rnglib_go

A Go random number generator especially useful for generating
quasi-random strings, data files, and directory trees containing
random numbers of data files with random names and random content.

## Classes

This package contains two classes, **SimpleRNG** and **SystemRNG**.
Both of these subclass Go's `rand` package and so `rand`'s functions
can be called through either of rnglib's two subclasses using
exactly the same syntax.

`SimpleRNG` is the faster of the two.  It uses the Mersenne
Twister and so is completely predictable with a very long period.
It is suitable where speed and predictability are both important.
You can be certain that if you provide the same seed, then
the sequence of numbers generated will be exactly the same.  This is
often important for debugging.

`SimpleRNG` is also very fast.  It is approximately 30% faster
than Go's rand package in our tests.

Although the Mersenne Twister is absolutely predictable, it also has
extremely good statistical properties.  Tests using standard
packages such as [Diehard](http://en.wikipedia.org/wiki/Diehard_tests)
and [Dieharder](http://www.phy.duke.edu/~rgb/General/dieharder.php) rank
the Mersenned Twister very highly.  The Twister is quite widely used
(for example, in the Python libraries) because of this.

`SystemRNG` is a secure random number generator, meaning that it is
extremely difficult or impossible to predict the next number
generated in a sequence.  It is based on the system's `/dev/urandom`.
This relies upon entropy accumulated by the system; when
this is insufficient it will revert to using a very secure
programmable random number generator.  `SystemRNG` is about 35x
slower than `SimpleRNG` in our test runs.

For normal use in testing, `SimpleRNG` is preferred.  When there is
a requirement for greater security, as in the generation of passwords
and cryptographic keys with reasonable strength, `SystemRNG` is recommended.

## Functions

In addition to the functions available from Go's random package,
rnglib also provides

	NextBoolean()
	NextByte(max=256)
	NextBytes(buffer)
	NextInt32(n uint32)
	NextInt64(n uint64)
	NextFloat32()
	NextFloat64()
	NextFileName(maxLen int)
	NextDataFile(dirName string, maxLen, minLen int)
	NextDataDir(pathToDir string, depth, width, maxLen, minLen int)

Given a buffer of arbitrary length, nextBytes() will fill it with random
bytes.

File names generated by `nextFileName()` are at least one character long.
The first letter must be alphabetic (including the underscore)
whereas other characters may also be numeric or either of dot and dash
("." and "-").

The function `nextDataFile()` creates a file with a random name in the
directory indicated.  The file length will vary from minLen (which
defaults to zero) up to but excluding maxLen bytes.

The `nextDataDir()` function creates a directory structure that is
'depth' deep and 'width' wide, where the latter means that there
will be that many files in each directory and subdirectory created
(where a file is either a data file or a directory). `depth` and
`width` must both be at least 1.  `maxLen` and `minLen` are used as in
`nextDataFile()`.

## History

A Java version of `rnglib` was originally implemented as part of the
[XLattice](http://www.xlattice.org)
project.  A Python implementation also exists: click
[here](https://jddixon.github.io/rnglib)
for further information`.

## Project Status

Stable.  All tests succeed, performance is good.

## On-line Documentation

More information on the **rnglib_go** project can be found
[here](https://jddixon.github.io/rnglib_go)


CHANGES
rnglib_go/CHANGES

v0.5.9
    2016-09-18
        * update LICENSE                                            SLOC 1638
v0.5.8
    2016-06-28
        * add ghpDoc/{css,img}                                      SLOC 1605
v0.5.7
    2016-06-01
        * add bkp2U, other bash scripts                             SLOC 1107
v0.5.6
    2016-02-25
        * update LICENSE                                            SLOC  608
v0.5.5
    2015-12-27
        * fix dead link
v0.5.4
    2015-11-18
        * minor edits to *.md                                       SLOC  654
v0.5.3
    2015-09-14
        * improved formatting, minor edits for index.html
v0.5.2
    2015-06-13
        * update .gitignore; add countem; minor cleanup             SLOC  654
v0.5.1
    2014-06-30
        * added SomeBytes(N)                                        SLOC  605
v0.5.0
    2014-05-10
        * split off from xlattice_go                                SLOC  600








prng_test.go
package rnglib_go

import (
	"fmt"
	xf "github.com/jddixon/xlUtil_go/lfs"
	. "gopkg.in/check.v1"
	"os"
	"strings"
	"testing"
	"time"
)

// SETUP FOR THIS PACKAGE (rnglib) //////////////////////////////////
const TMP_DIR = "tmp"

// gocheck tie-in /////////////////////
func Test(t *testing.T) { TestingT(t) }

type XLSuite struct{}

var _ = Suite(&XLSuite{})

// end gocheck setup //////////////////

// utility functions ////////////////////////////////////////////////
func (s *XLSuite) buildData(count uint32) []byte {
	p := make([]byte, count)
	return p
}
func (s *XLSuite) makeSimpleRNG() *PRNG {
	t := time.Now().Unix() // int64 sec
	rng := NewSimpleRNG(t)
	return rng
}

// parameterized unit tests /////////////////////////////////////////
func (s *XLSuite) doTestConstructor(c *C, rng *PRNG) {
	c.Assert(rng, Not(IsNil)) // NOT
}
func (s *XLSuite) doTestNextBoolean(c *C, rng *PRNG) {
	val := rng.NextBoolean()
	c.Assert(val, Not(IsNil)) // NOT

	valAsIface := interface{}(val)
	switch v := valAsIface.(type) {
	default:
		fmt.Printf("expected type bool, found %T", v)
		// assert.Fail("whatever NextBoolean() returns is not a bool")
	case bool:
		/* empty statement */
	}
}
func (s *XLSuite) doTestNextByte(c *C, rng *PRNG) {
	_ = c
	_ = rng
}
func (s *XLSuite) doTestNextBytes(c *C, rng *PRNG) {
	count := uint32(1)          // minimum length of buffer
	count += rng.NextInt32(256) // maximum
	data := s.buildData(count)  // so 1 .. 256 bytes
	rng.NextBytes(data)
	actualLen := uint32(len(data))
	c.Assert(0, Not(Equals), actualLen) // NOT
	c.Assert(actualLen, Equals, count)

}
func (s *XLSuite) doTestNextFileName(c *C, rng *PRNG) {
	maxLen := uint32(1)         // minimum length of name
	maxLen += rng.NextInt32(16) // maximum
	name := rng.NextFileName(int(maxLen))
	actualLen := len(name)
	c.Assert(0, Not(Equals), actualLen) // NOT
	// assert.True( t, actualLen < maxLen)
}
func (s *XLSuite) doTestNextDataFile(c *C, rng *PRNG) {
	minLen := int(rng.NextInt32(4))            // minimum length of file
	maxLen := minLen + int(rng.NextInt32(256)) // maximum

	// XXX should return err, which should be nil
	fileLen, pathToFile := rng.NextDataFile(TMP_DIR, maxLen, minLen)

	stats, err := os.Stat(pathToFile)
	c.Assert(err, IsNil)
	fileName := stats.Name()
	c.Assert(TMP_DIR+"/"+fileName, Equals, pathToFile)
	c.Assert(stats.Size(), Equals, int64(fileLen))

}
func (s *XLSuite) doNextDataDirTest(c *C, rng *PRNG, width int, depth int) {
	dirName := rng.NextFileName(8)
	dirPath := TMP_DIR + "/" + dirName
	pathExists, err := xf.PathExists(dirPath)
	if err != nil {
		panic("error invoking xf.PathExists on " + dirPath)
	}
	if pathExists {
		if strings.HasPrefix(dirPath, "/") {
			panic("attempt to remove absolute path " + dirPath)
		}
		if strings.Contains(dirPath, "..") {
			panic("attempt to remove path containing ..: " + dirPath)
		}
		os.RemoveAll(dirPath)
	}
	rng.NextDataDir(dirPath, width, depth, 32, 0)
}
func (s *XLSuite) doTestNextDataDir(c *C, rng *PRNG) {
	s.doNextDataDirTest(c, rng, 1, 1)
	s.doNextDataDirTest(c, rng, 1, 4)
	s.doNextDataDirTest(c, rng, 4, 1)
	s.doNextDataDirTest(c, rng, 4, 4)
}


bench_test.go
package rnglib_go

// to run, type "go test .bench ." remembering the dot

import (
	"math/rand"
	"testing"
	"time"
)

func BenchmarkRand(b *testing.B) {
	// typically 39.2 ns/op
	for i := 0; i < b.N; i++ {
		_ = rand.Int63() // Go's rand.Random package
	}
}

func BenchmarkSimpleRNG(b *testing.B) {
	t := time.Now().Unix()
	rng := NewSimpleRNG(t)
	b.ResetTimer()
	// typically 28.4 ns/op
	for i := 0; i < b.N; i++ {
		_ = rng.Int63() // Mersenne Twister
	}
}
func BenchmarkSystemRNG(b *testing.B) {
	t := time.Now().Unix()
	rng := NewSystemRNG(t)
	b.ResetTimer()
	// typically XX.X ns/op
	for i := 0; i < b.N; i++ {
		_ = rng.Int63() // /dev/urandom
	}
}


mt19937-64_test.go
package rnglib_go

// This is the original Hiroshima test hacked into Go -- and so
// doesn't actually do any useful unit tests.

import (
	"fmt"
	. "gopkg.in/check.v1"
)

func (s *XLSuite) TestMT64(c *C) {
	var (
		// no size, so it's a slice
		initX  = []uint64{0x12345, 0x23456, 0x34567, 0x45678}
		length = uint64(4)
	)
	m := NewMT64()
	m.init_by_array64(initX, length)
	fmt.Println("1000 outputs of genrand64_int64()")
	for i := 0; i < 1000; i++ {
		// fmt.Printf("%16v ", m.genrand64_int64())
		fmt.Printf("%20v ", m.genrand64_int64())
		if i%5 == 4 {
			fmt.Printf("\n")
		}
	}
	fmt.Printf("\n1000 outputs of genrand64_real2()\n")
	for i := 0; i < 1000; i++ {
		fmt.Printf("%10.8f ", m.genrand64_real2())
		if i%5 == 4 {
			fmt.Printf("\n")
		}
	}
}


mt19937-64.go
// mt19937-64 with license removed for brevity

package rnglib_go

const NN int = 312
const MM int = 156
const MATRIX_A uint64 = 0xB5026F5AA96619E9
const UM uint64 = 0xFFFFFFFF80000000 /* Most significant 33 bits */
const LM uint64 = 0x7FFFFFFF         /* Least significant 31 bits */

type MT64 struct {
	mt  [NN]uint64 // state vector
	mti int
}

func NewMT64() *MT64 {
	return &MT64{mti: NN + 1} // means mt[NN] is not initialized
}

/* initializes mt[NN] with a seed */
func (m *MT64) init_genrand64(seed uint64) {
	m.mt[0] = seed
	for m.mti = 1; m.mti < NN; m.mti++ {
		m.mt[m.mti] = uint64(6364136223846793005)*(m.mt[m.mti-1]^(m.mt[m.mti-1]>>62)) + uint64(m.mti)
	}
}

/* initialize by an array with array-length */
/* init_key is the array for initializing keys */
/* key_length is its length */
func (m *MT64) init_by_array64(init_key []uint64, key_length uint64) {
	var i, j, k uint64
	const NN64 = uint64(NN)
	m.init_genrand64(uint64(19650218))
	i = 1
	j = 0
	if NN64 > key_length {
		k = NN64
	} else {
		k = key_length
	}
	for ; k > 0; k-- {
		m.mt[i] = (m.mt[i] ^ ((m.mt[i-1] ^ (m.mt[i-1] >> 62)) * 3935559000370003845)) + init_key[j] + j /* non linear */
		i++
		j++
		if i >= NN64 {
			m.mt[0] = m.mt[NN-1]
			i = 1
		}
		if j >= key_length {
			j = 0
		}
	}
	for k = NN64 - 1; k > 0; k-- {
		m.mt[i] = (m.mt[i] ^ ((m.mt[i-1] ^ (m.mt[i-1] >> 62)) * 2862933555777941757)) - i /* non linear */
		i++
		if i >= NN64 {
			m.mt[0] = m.mt[NN64-1]
			i = 1
		}
	}

	m.mt[0] = uint64(1) << 63 // MSB is 1; assuring non-zero initial array
}

/* generates a random number on [0, 2^64-1]-interval */
func (m *MT64) genrand64_int64() uint64 {
	var (
		i int
		x uint64
	)
	var mag01 = [2]uint64{0, MATRIX_A}
	// const NN64 = uint64(NN)

	if m.mti >= NN { /* generate NN words at one time */

		/* if init_genrand64() has not been called, */
		/* a default initial seed is used     */
		if m.mti == NN+1 {
			m.init_genrand64(uint64(5489))
		}

		for i = 0; i < NN-MM; i++ {
			x = (m.mt[i] & UM) | (m.mt[i+1] & LM)
			m.mt[i] = m.mt[i+MM] ^ (x >> 1) ^ mag01[int((x&uint64(1)))]
		}
		for ; i < NN-1; i++ {
			x = (m.mt[i] & UM) | (m.mt[i+1] & LM)
			m.mt[i] = m.mt[i+(MM-NN)] ^ (x >> 1) ^ mag01[(int)(x&uint64(1))]
		}
		x = (m.mt[NN-1] & UM) | (m.mt[0] & LM)
		m.mt[NN-1] = m.mt[MM-1] ^ (x >> 1) ^ mag01[(int)(x&uint64(1))]

		m.mti = 0
	}
	x = m.mt[m.mti]
	m.mti++

	x ^= (x >> 29) & uint64(0x5555555555555555)
	x ^= (x << 17) & uint64(0x71D67FFFEDA60000)
	x ^= (x << 37) & uint64(0xFFF7EEE000000000)
	x ^= (x >> 43)

	return x
}

/* generates a random number on [0, 2^63-1]-interval */
func (m *MT64) genrand64_int63() uint64 {
	return uint64(m.genrand64_int64() >> 1)
}

/* generates a random number on [0,1]-real-interval */
func (m *MT64) genrand64_real1() float64 {
	return float64(m.genrand64_int64()>>11) * (1.0 / 9007199254740991.0)
}

/* generates a random number on [0,1)-real-interval */
func (m *MT64) genrand64_real2() float64 {
	return float64(m.genrand64_int64()>>11) * (1.0 / 9007199254740992.0)
}

/* generates a random number on (0,1)-real-interval */
func (m *MT64) genrand64_real3() float64 {
	return (float64(m.genrand64_int64()>>12) + 0.5) * (1.0 / 4503599627370496.0)
}

// rand.Source interface ////////////////////////////////////////////

func (m *MT64) Seed(seed int64) {
	m.init_genrand64(uint64(seed))
}
func (m *MT64) Int63() int64 {
	return int64(m.genrand64_int63())
}


system_rng.go
package rnglib_go

import (
	"bufio"
	"encoding/binary"
	"io"
	"math/rand"
	"os"
	"sync"
)

// SystemRNG draws values from /dev/urandom and runs about 35x slower
// than SimpleRNG, which relies upon the 64-bit Mersenne Twister.  Both
// share the same interface, including Go's math.rand functions.
//
// SystemRNG is a reasonably secure random number generator.  It
// ignores any seed provided.  On Linux, if you need a more secure
// source of random values, you can read /dev/random, but this will
// block if there is not enough entropy available.

type uReader struct {
	name string
	f    io.Reader
	mu   sync.Mutex
}

func (r *uReader) Read(b []byte) (n int, err error) {
	// with locking, about 1050ns/op; without, about 900ns/op
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.f == nil {
		f, err := os.Open(r.name)
		if f == nil {
			return 0, err
		}
		r.f = bufio.NewReader(f)
	}
	return r.f.Read(b)
}

// rand.Source interface ////////////////////////////////////////////

func (r *uReader) Seed(seed int64) {
	_ = seed
}
func (r *uReader) Int63() int64 {
	var n uint64
	// Given that this is random data, it doesn't really matter
	// whether we regard it as big- or little-endian, and so we
	// should choose whichever does NOT result in bytes being
	// reordered on the current host.
	err := binary.Read(r, binary.LittleEndian, &n)
	if err != nil {
		panic("error reading from /dev/urandom")
	}
	val := int64(n >> 1)
	return val
}

// SystemRNG ////////////////////////////////////////////////////////

func NewSystemSource() rand.Source {
	var dReader uReader
	dReader.name = "/dev/urandom"
	return &dReader
}
func NewSystemRNG(seed int64) *PRNG {
	s := new(PRNG)
	src := NewSystemSource()
	s.rng = rand.New(src)
	return s
}


prng.go
package rnglib_go

import (
	xf "github.com/jddixon/xlUtil_go/lfs"
	"math/rand"
	"os"
	"strings"
)

type PRNG struct {
	rng *rand.Rand
}

//func NewMTSource(seed int64) rand.Source {
//	var mt64 MT64
//	mt64.Seed(seed)
//	return &mt64
//}
//
//func NewSimpleRNG(seed int64) *PRNG {
//	p := new(PRNG) // allocates
//	src := NewMTSource(seed)
//	p.rng = rand.New(src)
//	p.Seed(seed)
//	return s
//}

func (p *PRNG) Seed(seed int64) {
	p.rng.Seed(seed)
}

// expose the rand.Random interface
func (p *PRNG) Int63() int64         { return p.rng.Int63() }
func (p *PRNG) Uint32() uint32       { return p.rng.Uint32() }
func (p *PRNG) Int31() int32         { return p.rng.Int31() }
func (p *PRNG) Int() int             { return p.rng.Int() }
func (p *PRNG) Int63n(n int64) int64 { return p.rng.Int63n(n) }
func (p *PRNG) Int31n(n int32) int32 { return p.rng.Int31n(n) }
func (p *PRNG) Intn(n int) int       { return p.rng.Intn(n) }
func (p *PRNG) Float64() float64     { return p.rng.Float64() }
func (p *PRNG) Float32() float32     { return p.rng.Float32() }
func (p *PRNG) Perm(n int) []int     { return p.rng.Perm(n) }

// PRNG functions
func (p *PRNG) NextBoolean() bool { return p.rng.Int63()&1 == 0 }
func (p *PRNG) NextByte() byte    { return byte(p.rng.Int63n(256)) }

// miraculously inefficient
func (p *PRNG) NextBytes(buffer []byte) {
	for n := range buffer {
		buffer[n] = p.NextByte()
	}
}

// Allocate a slice of N bytes, fill with random values, return.
func (p *PRNG) SomeBytes(n uint) []byte {
	b := make([]byte, n)
	p.NextBytes(b)
	return b
}
func (p *PRNG) NextInt32(n uint32) uint32 { return uint32(float32(n) * p.rng.Float32()) }
func (p *PRNG) NextInt64(n uint64) uint64 { return uint64(float64(n) * p.rng.Float64()) }
func (p *PRNG) NextFloat32() float32      { return p.rng.Float32() }
func (p *PRNG) NextFloat64() float64      { return p.rng.Float64() }

// These produce strings which are acceptable POSIX file namep.
// and also advance a cursor by a multiple of 64 bitp.  All strings
// are at least one byte and less than maxLen bytes n length.  We
// arbitrarily limit file names to less than 256 characterp.
func (p *PRNG) _nextFileName(nameLen int) string {
	/* always returns at least one character */
	maxStarterNdx := uint32(len(_FILE_NAME_STARTERS))
	ndx := p.NextInt32(maxStarterNdx)
	var chars []string
	chars = append(chars, _FILE_NAME_STARTERS[ndx])
	maxCharNdx := uint32(len(_FILE_NAME_CHARS))
	for n := 0; n < nameLen; n++ {
		ndx := p.NextInt32(maxCharNdx)
		chars = append(chars, _FILE_NAME_CHARS[ndx])
	}
	return strings.Join(chars, "")
}

func (p *PRNG) NextFileName(maxLen int) string {
	_maxLen := uint32(maxLen)
	if _maxLen < 2 {
		_maxLen = 2 // this is a ceiling which cannot be reached
	}
	if _maxLen > 256 {
		_maxLen = 256
	}
	var name string
	nameLen := uint32(0)
	for nameLen == 0 {
		nameLen = p.NextInt32(_maxLen) // so len < 256
	}
	for {
		name = p._nextFileName(int(nameLen))
		if (len(name) > 0) && !strings.Contains(name, "..") {
			break
		}
	}
	return name
}

// Return a string which is an acceptable POSIX path with at least
// minParts parts and fewer than maxParts parts.  If the parameters
// supplied are unreasonable, they are adjusted.
//
// If uniqueness or some other constraint needs to be met, call this
// function then test compliance, then as necessary repeat.
func (p *PRNG) NextPosixPath(maxPartLen, minParts, maxParts int) (pp string) {

	if maxParts < 2 {
		maxParts = 2 // so at least one
	}
	if minParts <= 0 || maxParts <= minParts {
		minParts = maxParts - 1
	}
	count := minParts + p.Intn(maxParts-minParts)
	var parts []string
	for i := 0; i < count; i++ {
		parts = append(parts, p.NextFileName(maxPartLen))
	}
	return strings.Join(parts, "/")
}

// Return a string which is a valid fully qualified domain name.  Only
// names in .com, .net, and .org are returned by this version.  The
// parameter is a suggestion as to the complexity of the result, with
// zero returning the simplest response and values greater than zero
// possibly returning a more complex result.
func (p *PRNG) NextFQDN(complexity int) string {

	if complexity == 0 {
		complexity = 0
	} else if complexity > 3 {
		complexity = 3
	}
	top := p.Intn(3)
	var topLevel string
	switch top {
	case 0:
		topLevel = "com"
	case 1:
		topLevel = "net"
	default:
		topLevel = "org"
	}
	var parts []string
	count := 1 + complexity
	for i := 0; i < count; i++ {
		parts = append(parts, p.NextFileName(8))
	}
	parts = append(parts, topLevel)
	return strings.Join(parts, ".")
}

// Return a string which is a well-formed email address.  Complexity
// is a suggestion as to how complex the result should be.
func (p *PRNG) NextEmailAddress(complexity int) (e string) {
	if complexity == 0 {
		complexity = 0
	} else if complexity > 3 {
		complexity = 3
	}
	fqdn := p.NextFQDN(complexity)
	name := p.NextFileName(5 + complexity)
	return name + "@" + fqdn
}

// OPERATIONS ON THE FILE SYSTEM ------------------------------------

// These are operations on the file system.  Directory depth is at least 1
// and no more than 'depth'.  Likewise for width, the number of
// files in a directory, where a file is either a data file or a subdirectory.
// The number of bytes in a file is at least minLen and less than maxLen.
// Subdirectory names may be random
//
// XXX Changed to return an int64 length.
//
func (p *PRNG) NextDataFile(dirName string, maxLen int, minLen int) (
	int64, string) {

	// silently convert parameters to reasonable values
	if minLen < 0 {
		minLen = 0
	}
	if maxLen < minLen+1 {
		maxLen = minLen + 1
	}

	// create the data directory if it does not exist
	dirExists, err := xf.PathExists(dirName)
	if err != nil {
		panic(err)
	}
	if !dirExists {
		os.MkdirAll(dirName, 0755)
	}

	// loop until the file does not exist
	pathToFile := dirName + "/" + p.NextFileName(16)
	pathExists, err := xf.PathExists(pathToFile)
	if err != nil {
		panic(err)
	}
	for pathExists {
		pathToFile := dirName + "/" + p.NextFileName(16)
		pathExists, err = xf.PathExists(pathToFile)
		if err != nil {
			panic(err)
		}
	}
	count := minLen + int(p.NextFloat32()*float32((maxLen-minLen)))
	data := make([]byte, count)
	p.NextBytes(data) // fill with random bytes
	fo, err := os.Create(pathToFile)
	if err != nil {
		panic(err)
	}
	defer func() {
		if err := fo.Close(); err != nil {
			panic(err)
		}
	}() // XXX wakaranai
	// XXX this should be chunked
	// XXX data should be slice
	if _, err := fo.Write(data); err != nil {
		panic(err)
	}
	// XXX respec to also return err
	return int64(count), pathToFile
}

// NextDataDir creates a directory tree populated with data filep.
//
// BUGS
// * on at least one occasion with width = 4 only 3 files/directories
//   were created at the top level (2 were subdirs)
// DEFICIENCIES:
// * no control over percentage of directories
// * no guarantee that depth will be reached
//
func (p *PRNG) NextDataDir(pathToDir string, depth int, width int,
	maxLen int, minLen int) {
	// number of directory levels; 1 means no subdirectories
	if depth < 1 {
		depth = 1
	}
	// number of members (files, subdirectories) at each level
	if width < 1 {
		width = 1
	}
	// XXX may panic
	pathExists, err := xf.PathExists(pathToDir)
	if err != nil {
		panic(err)
	}
	if !pathExists {
		os.MkdirAll(pathToDir, 0755)
	}
	subdirSoFar := 0
	for i := 0; i < width; i++ {
		if depth > 1 {
			if (p.NextFloat32() > 0.25) &&
				((i < width-1) || (subdirSoFar > 0)) {
				// 25% are subdirs
				// data file i
				// SPECIFICATION ERROR: file name may not be unique
				// count, pathToFile
				p.NextDataFile(pathToDir,
					maxLen, minLen)
			} else {
				// directory
				subdirSoFar += 1
				// create unique name
				fileName := p.NextFileName(16)
				pathToSubdir := pathToDir + "/" + fileName
				p.NextDataDir(pathToSubdir, depth-1, width,
					maxLen, minLen)
			}
		} else {
			// data file
			// XXX SPECIFICATION ERROR: file name may not be unique
			// count, pathToFile
			p.NextDataFile(pathToDir, maxLen, minLen)
		}
	}
}
